unrouted_handler.go

```diff
@@ -437,12 +450,27 @@ func (handler *UnroutedHandler) PatchFile(w http.ResponseWriter, r *http.Request
 	}
 
 	// Do not proxy the call to the data store if the upload is already completed
-	if info.Offset == info.Size {
+	if !info.SizeIsDeferred && info.Offset == info.Size {
 		w.Header().Set("Upload-Offset", strconv.FormatInt(offset, 10))
 		handler.sendResp(w, r, http.StatusNoContent)
 		return
 	}
 
+	if handler.composer.UsesLengthDeferrer && info.SizeIsDeferred && r.Header.Get("Upload-Length") != "" {
+		uploadLength, err := strconv.ParseInt(r.Header.Get("Upload-Length"), 10, 64)
+		if err != nil || uploadLength < 0 {
+			handler.sendError(w, r, ErrInvalidOffset)
+			return
+		}
+
+		info.Size = uploadLength
+		info.SizeIsDeferred = false
+		if err := handler.composer.LengthDeferrer.DeclareLength(id, info.Size); err != nil {
```

Comment:
An additional check if the new length is not smaller than the current offset (`info.Offset <= uploadLength`) and the new length does not exceed the file limit (see https://github.com/tus/tusd/blob/master/unrouted_handler.go#L264-L268) before the call to `DeclareLength` would be needed.